File: CoreScopeViewer/Features/Onboarding/OnboardingScreen.swift
import SwiftUI

struct OnboardingScreen: View {
    @Environment(AnalyzerSourceRegistry.self) private var sourceRegistry
    @Environment(AnalyzerSettings.self) private var settings
    let onComplete: () -> Void

    @State private var selectedHost = ""
    @State private var showsSourcePicker = false
    @State private var step: Step = .source

    private enum Step {
        case source
        case overview
    }

    var body: some View {
        NavigationStack {
            Group {
                switch step {
                case .source:
                    sourceSelection
                case .overview:
                    overview
                }
            }
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                if step == .overview {
                    ToolbarItem(placement: .topBarLeading) {
                        Button("Back") {
                            withAnimation {
                                step = .source
                            }
                        }
                    }
                }
            }
            .sheet(isPresented: $showsSourcePicker) {
                NavigationStack {
                    AnalyzerSourcePickerScreen(
                        selectedHost: $selectedHost,
                        appliesSelectionImmediately: true,
                        onSelection: nil
                    )
                }
            }
            .task {
                if selectedHost.isEmpty {
                    selectedHost = settings.host
                }
                await sourceRegistry.refresh()
                if selectedHost.isEmpty {
                    selectedHost = sourceRegistry.defaultSource?.host ?? ""
                }
            }
        }
    }

    private var sourceSelection: some View {
        VStack(spacing: 24) {
            Spacer()

            Image(systemName: "antenna.radiowaves.left.and.right.circle.fill")
                .font(.system(size: 76))
                .foregroundStyle(Color.accentColor)

            VStack(spacing: 10) {
                Text("Welcome to NodeScope")
                    .font(.largeTitle.weight(.bold))
                Text("Choose a community analyzer to explore its live nodes, routes, channels, and observers.")
                    .font(.body)
                    .foregroundStyle(Color.secondary)
                    .multilineTextAlignment(.center)
                    .padding(.horizontal, 28)
            }

            Button {
                showsSourcePicker = true
            } label: {
                HStack(spacing: 12) {
                    Image(systemName: "antenna.radiowaves.left.and.right")
                        .foregroundStyle(Color.accentColor)
                    VStack(alignment: .leading, spacing: 2) {
                        Text(selectedSourceName)
                            .font(.headline)
                            .foregroundStyle(Color.primary)
                        Text(selectedHost.isEmpty ? "Choose an analyzer" : selectedHost)
                            .font(.caption)
                            .foregroundStyle(Color.secondary)
                    }
                    Spacer()
                    Image(systemName: "chevron.right")
                        .font(.caption.weight(.semibold))
                        .foregroundStyle(Color.secondary)
                }
                .padding(16)
                .frame(maxWidth: .infinity, alignment: .leading)
                .contentShape(Rectangle())
                .background(.thinMaterial, in: RoundedRectangle(cornerRadius: 18, style: .continuous))
            }
            .buttonStyle(.plain)
            .padding(.horizontal, 24)
            .accessibilityLabel("Analyzer source: \(selectedSourceName)")

            Text("You can change sources anytime in Settings, or add a custom analyzer.")
                .font(.caption)
                .foregroundStyle(Color.secondary)
                .multilineTextAlignment(.center)
                .padding(.horizontal, 32)

            Spacer()

            Button("Continue") {
                withAnimation {
                    step = .overview
                }
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.large)
            .disabled(selectedHost.isEmpty)
            .padding(.bottom, 28)
        }
        .adaptiveContentWidth(720)
    }

    private var overview: some View {
        VStack(spacing: 24) {
            Spacer()

            Image(systemName: "map.fill")
                .font(.system(size: 64))
                .foregroundStyle(Color.accentColor)

            VStack(spacing: 8) {
                Text("Follow the Mesh")
                    .font(.largeTitle.weight(.bold))
                Text("NodeScope makes live activity easier to understand at a glance.")
                    .font(.body)
                    .foregroundStyle(Color.secondary)
                    .multilineTextAlignment(.center)
                    .padding(.horizontal, 28)
            }

            VStack(alignment: .leading, spacing: 16) {
                OnboardingFeatureRow(
                    icon: "circle.fill",
                    title: "Explore nodes",
                    detail: "Tap a node for its details. Clusters zoom in to reveal nearby nodes."
                )
                OnboardingFeatureRow(
                    icon: "point.topleft.down.curvedto.point.bottomright.up",
                    title: "Watch packet routes",
                    detail: "Live routes animate hop by hop. Open a channel message to replay its path."
                )
                OnboardingFeatureRow(
                    icon: "line.3.horizontal.decrease.circle",
                    title: "Focus your view",
                    detail: "Choose a region on the map to narrow traffic and center the view."
                )
            }
            .padding(18)
            .background(.thinMaterial, in: RoundedRectangle(cornerRadius: 20, style: .continuous))
            .padding(.horizontal, 24)

            Text("Location is optional and is only used when you choose to center the map on yourself.")
                .font(.caption)
                .foregroundStyle(Color.secondary)
                .multilineTextAlignment(.center)
                .padding(.horizontal, 32)

            Spacer()

            Button("Start Exploring") {
                settings.host = selectedHost
                onComplete()
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.large)
            .padding(.bottom, 28)
        }
        .adaptiveContentWidth(720)
    }

    private var selectedSourceName: String {
        sourceRegistry.sources.first {
            $0.host.caseInsensitiveCompare(selectedHost) == .orderedSame
        }?.name ?? "Custom Analyzer"
    }
}

private struct OnboardingFeatureRow: View {
    let icon: String
    let title: String
    let detail: String

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            Image(systemName: icon)
                .font(.body)
                .foregroundStyle(Color.accentColor)
                .frame(width: 22)
                .padding(.top, 2)
            VStack(alignment: .leading, spacing: 3) {
                Text(title)
                    .font(.subheadline.weight(.semibold))
                    .foregroundStyle(Color.primary)
                Text(detail)
                    .font(.caption)
                    .foregroundStyle(Color.secondary)
            }
        }
    }
}

File: android/app/src/main/java/org/nodescope/android/feature/onboarding/OnboardingScreen.kt
package org.nodescope.android.feature.onboarding

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material.icons.outlined.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import org.nodescope.android.R
import org.nodescope.android.core.model.AnalyzerSource
import org.nodescope.android.core.network.normalizeHost
import org.nodescope.android.core.storage.SourceIcons

private enum class OnboardingStep { SOURCE, OVERVIEW }

/**
 * First launch, as on iOS (`OnboardingScreen`): choose an analyzer, then a short overview of
 * what the app shows, then "Start exploring" saves the source. The analyzer card opens the
 * same picker Settings uses; choosing there comes back here with that source selected.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun OnboardingScreen(
    sources: List<AnalyzerSource>, initialHost: String, saving: Boolean, error: String?,
    onFinish: (String) -> Unit, icons: SourceIcons? = null,
) {
    var step by rememberSaveable { mutableStateOf(OnboardingStep.SOURCE) }
    var host by rememberSaveable { mutableStateOf(initialHost) }
    var picking by rememberSaveable { mutableStateOf(false) }
    var pickError by remember { mutableStateOf<String?>(null) }
    BackHandler(enabled = picking || step == OnboardingStep.OVERVIEW) {
        if (picking) picking = false else step = OnboardingStep.SOURCE
    }
    if (picking) {
        Column(Modifier.fillMaxSize()) {
            TopAppBar(title = { Text("Analyzer source") }, windowInsets = WindowInsets(0, 0, 0, 0), navigationIcon = {
                IconButton(onClick = { picking = false }) { Icon(Icons.AutoMirrored.Outlined.ArrowBack, "Back") }
            })
            SourceScreen(sources, host, saving = false, error = pickError, onSave = { chosen ->
                runCatching { normalizeHost(chosen) }
                    .onSuccess { host = it; pickError = null; picking = false }
                    .onFailure { pickError = it.message ?: "Enter a valid analyzer hostname." }
            }, icons = icons)
        }
        return
    }
    val source = sources.firstOrNull { runCatching { normalizeHost(it.host) }.getOrNull().equals(host, true) }
    when (step) {
        OnboardingStep.SOURCE -> OnboardingPage(
            icon = Icons.Outlined.SettingsInputAntenna, title = stringResource(R.string.welcome),
            description = "Choose a community analyzer to explore its live nodes, routes, channels, and observers.",
            primary = stringResource(R.string.continue_label), primaryEnabled = host.isNotBlank(),
            onPrimary = { step = OnboardingStep.OVERVIEW },
        ) {
            Card(onClick = { picking = true }, modifier = Modifier.fillMaxWidth()) {
                Row(Modifier.padding(16.dp), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                    if (source != null) SourceLogo(source, icons)
                    else Icon(Icons.Outlined.SettingsInputAntenna, null, Modifier.size(40.dp).padding(8.dp), tint = MaterialTheme.colorScheme.primary)
                    Column(Modifier.weight(1f)) {
                        Text(source?.name ?: "Custom analyzer", style = MaterialTheme.typography.titleMedium)
                        Text(host.ifBlank { "Choose an analyzer" }, style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant)
                    }
                    Icon(Icons.Outlined.ChevronRight, "Choose analyzer", tint = MaterialTheme.colorScheme.onSurfaceVariant)
                }
            }
            Text("You can change sources anytime in Settings, or add a custom analyzer.", textAlign = TextAlign.Center,
                style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }
        OnboardingStep.OVERVIEW -> OnboardingPage(
            icon = Icons.Outlined.Map, title = "Follow the mesh",
            description = "NodeScope makes live activity easier to understand at a glance.",
            primary = stringResource(if (saving) R.string.saving else R.string.start_exploring), primaryEnabled = !saving,
            onPrimary = { onFinish(host) }, onBack = { step = OnboardingStep.SOURCE },
        ) {
            Card(Modifier.fillMaxWidth()) {
                Column(Modifier.padding(18.dp), verticalArrangement = Arrangement.spacedBy(16.dp)) {
                    FeatureRow(Icons.Outlined.Hub, "Explore nodes", "Tap a node for its details. Clusters zoom in to reveal nearby nodes.")
                    FeatureRow(Icons.Outlined.Route, "Watch packet routes",
                        "Live routes animate hop by hop. Open a channel message to replay its path.")
                    FeatureRow(Icons.Outlined.FilterList, "Focus your view", "Choose a region on the map to narrow traffic and center the view.")
                }
            }
            Text("Location is optional and is only used when you choose to center the map on yourself.", textAlign = TextAlign.Center,
                style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            if (error != null) Text(error, color = MaterialTheme.colorScheme.error, textAlign = TextAlign.Center)
        }
    }
}

/** Icon, title and description, the step's content, then the primary action at the bottom. */
@Composable
private fun OnboardingPage(
    icon: ImageVector, title: String, description: String, primary: String, primaryEnabled: Boolean,
    onPrimary: () -> Unit, onBack: (() -> Unit)? = null, content: @Composable ColumnScope.() -> Unit,
) {
    Column(Modifier.fillMaxSize()) {
        Box(Modifier.fillMaxWidth().height(56.dp)) {
            onBack?.let { back -> TextButton(onClick = back, Modifier.align(Alignment.CenterStart).padding(start = 8.dp)) { Text("Back") } }
        }
        // Centered when it fits; scrolls (e.g. at large text sizes) when it doesn't.
        BoxWithConstraints(Modifier.weight(1f).fillMaxWidth()) {
            Column(Modifier.fillMaxWidth().verticalScroll(rememberScrollState()).heightIn(min = maxHeight).padding(horizontal = 24.dp),
                horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.Center) {
                Column(Modifier.widthIn(max = 720.dp), horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(20.dp)) {
                    Icon(icon, null, Modifier.size(72.dp), tint = MaterialTheme.colorScheme.primary)
                    Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        Text(title, Modifier.semantics { heading() }, style = MaterialTheme.typography.headlineLarge,
                            fontWeight = FontWeight.Bold, textAlign = TextAlign.Center)
                        Text(description, textAlign = TextAlign.Center, color = MaterialTheme.colorScheme.onSurfaceVariant)
                    }
                    content()
                }
            }
        }
        Button(onClick = onPrimary, enabled = primaryEnabled,
            modifier = Modifier.padding(24.dp).widthIn(max = 480.dp).fillMaxWidth().align(Alignment.CenterHorizontally)) { Text(primary) }
    }
}

@Composable
private fun FeatureRow(icon: ImageVector, title: String, detail: String) {
    Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
        Icon(icon, null, Modifier.padding(top = 2.dp).size(22.dp), tint = MaterialTheme.colorScheme.primary)
        Column(verticalArrangement = Arrangement.spacedBy(3.dp)) {
            Text(title, style = MaterialTheme.typography.titleSmall)
            Text(detail, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }
    }
}
